### ISSUE TITLE
corruption in rendering in both openCL on and off with multiple color balance rgb modules

### ISSUE THREAD

**Issue by garrett** [darktable-org/darktable#14393]
### Describe the bug

I was working on a "play raw" file from the Pixls.us forum, from https://discuss.pixls.us/t/stairway-to-heaven/36881, using a copy of the color calibration module with a mask (and inverted mask on the second copy) to color balance for two parts of the image, as outlined in https://www.youtube.com/watch?v=U4CEN0JPcoM, and hit a corruption bug. The only way to work around the issue was to restart darktable. Toggling openCL off didn't help.

![Screenshot from 2023-05-01 17-31-53](https://user-images.githubusercontent.com/10246/235482038-cebc6142-0437-459e-b0f6-9ada329381c4.png)

### Steps to reproduce

1. download https://discuss.pixls.us/uploads/short-url/6SNZVfz4ahddbBWva55wFYEzPHM.RAF
2. use this XMP: https://discuss.pixls.us/uploads/short-url/4x2mSONurioOrVpglqCwjyJGbCW.xmp (make sure it's called the same as the filename — both _should_ be named properly when downloaded, but double-check)
3. open it in darktable and adjust some settings... a quick way is to adjust the color balance in one of the color calibration modules
4. see corruption in the main image, but not the thumbnail! (probably!)
5. restart if you want to get back to a normal state

Other images have been fine, but I _normally_ don't do split color balance with two copies of the color calibration module.

### Expected behavior

darktable should not corrupt the image while working on it

### Logfile | Screenshot | Screencast

The image should keep looking like this: 

![DSCF2164_12](https://user-images.githubusercontent.com/10246/235482757-0a546b04-d781-41a7-8cd3-b0fa3d0aeea2.jpg)

Not like this:

![image](https://user-images.githubusercontent.com/10246/235482877-7b558399-6ce3-43aa-851c-5e72297a59b7.png)

### Where did you install darktable from?

OSB

### darktable version

4.3.0~git2075.b040bade-9910.1

### What OS are you using?

Linux

### What is the version of your OS?

Fedora 38 Silverblue

### Describe your system?

AMD Ryzen 5 5600X
AMD RX 6700XT GPU
32 GiB RAM
X11 (for color management support)
GNOME 44.1
GTK 3.24.37

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

AMD RX 6700XT GPU, rocm-opencl-5.4.3-2.fc38.x86_64

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

I have tried openCL both on and off (with restarts in-between, to be sure) and the corruption shows up in both.

I'm not sure the problem is related to having two color balance modules (with a mask in the first and shared with the second as a reversed raster mask of the first), or if it is limited to that (as in, if it might affect other photos... if this is a mask issue, for example).

Previous versions of darktable used to work fine; but I haven't tried this technique in a while, so I don't know where it showed up.  Aurélien's video @ https://www.youtube.com/watch?v=U4CEN0JPcoM shows the technique there and it clearly works. However, I think he wasn't using a raster mask, but reusing a drawn mask?

**Comment by jenshannoschwalm** [comment]
Any raster mask involved while developing?

**Comment by garrett** [comment]
Yes, I was using a raster mask on the second color calibration module; it was based on the first color calibration module's parametric mask (so they would work on separate parts of the image). Adjusting color calibration triggered the issue.

**Comment by jenshannoschwalm** [comment]
Thanks for reporting :-) (Sorry i missed the raster mask mentioned in the original report btw)

This might be related to a late regression (reported as #14398) Also we identified some issues related to rsater masks as reported in #14271 and a pr currently under discussion #14376.

Let's see and reinvestigate later ...

**Comment by TurboGit** [comment]
@garrett : Can you test with current master? I have reverted some code identified as wrong. TIA.

**Comment by garrett** [comment]
Sure thing! I'll try to test it tonight.

(I'm using the OBS package on Fedora Silverblue, I can also try building master from git again inside a toolbox.)

**Comment by TurboGit** [comment]
Hum, looks like another issue. Can you try to fully revert the following commit, compile & check again:

`$ git revert a5ee7e75f7d50e749fb88109d21bce1cf0b4abf1`

TIA.

**Comment by jenshannoschwalm** [comment]
Two things,

1. you are using lens twice. That a) doesn't make sense for me as the lens parameters apply to whole-image-data. I am even not sure if the roi expanding code works correctly here and keeps data in allowed limits.. We should allow this module applied only once imho.
2. While zoomin in&out i saw the issue once. Still - the output seems to vary "a lot" depending on scale / zoom of the darkroom window.

**Comment by garrett** [comment]
> you are using lens twice

I noticed that too. It's showing up on fresh edits for me, regardless of the camera. I think it's a bug in darktable master? I meant to file it, but forgot about it.

**Comment by Nilvus** [comment]
> > you are using lens twice
> 
> I noticed that too. It's showing up on fresh edits for me, regardless of the camera. I think it's a bug in darktable master? I meant to file it, but forgot about it.

Check your presets before. I have presets and use master but never see 2 lens modules. Could be a bug in master but could also be on your side.

**Comment by jenshannoschwalm** [comment]
The  lens module die not have a flag to prohibit that yet.

**Comment by garrett** [comment]
> Hum, looks like another issue. Can you try to fully revert the following commit, compile & check again:
> $ git revert a5ee7e75f7d50e749fb88109d21bce1cf0b4abf1

I did the revert and still have the same issue. It doesn't _usually_ immediately happen, but it did happen after changing the hue a few times.

---

As for the lens correct modules being duplicated, when I went to the path where I have pixls.us images from playraws and run `darktable --configdir darktable --library darktable.db .` I don't see the module duplicated.

After moving files around, I found it was in my `data.db`.

Poking around further, I did have a preset to automatically apply a lens profile, which worked in prior versions of darktable.

![image](https://user-images.githubusercontent.com/10246/235774250-7430e10e-952b-441a-9086-f49a57f8b423.png)

After removing that preset and resetting some images, I see some images from different cameras over the years have lens profiles automatically applied and others do not... I'm not sure why it's sometimes applied and sometimes not. Most of the cameras+lenses I have are profiled. A few ancient ones should be, but aren't.

Removing the preset did fix the double module & double apply. Thanks! I created that preset because I always want lens correction applied by default... and it previously worked. Now? 

I'm not sure how to always have lens correction on without this bug. :frowning_face: (I almost always want both lens correction and noise removal on... if there's a matching profile that works for both.) Shouldn't darktable have defaults for these to always enable the modules if there's a matching profile? I guess I should open an enhancement request...

**Comment by TurboGit** [comment]
> The lens module die not have a flag to prohibit that yet.

Support for multiple lens instance has been added 2 years ago:

```
commit 47ec421cdce76fb9345625c363f5e0734f0fd7ce
Author: Chris.Elston <chris.elston@blueyonder.co.uk>
Date:   Sat Apr 17 20:59:28 2021 +0100

    lens: allow multi-instance
    
    lens correction can be run in both "correct" and "distort" modes
    
    however, in order to simulate another lens with the "distort" mode, the
    image (presumably) first has to be corrected, which requires a second
    instance
```

**Comment by TurboGit** [comment]
> Removing the preset did fix the double module & double apply. Thanks! I created that preset because I always want lens correction applied by default... and it previously worked. Now?

There was a bug (feature?) in dt where only the first preset for a given module was applied. This has been fixed in 4.3, so probably why you see this issue now.

Yet, there is probably a bug somewhere as we do allow now multiple lens instance.

**Comment by garrett** [comment]
I found a previous discussion about having darktable enable lens profiles by default:

- https://github.com/darktable-org/darktable/issues/13408

I haven't found one for noise, but I guess the PR linked from the above issue is kind of related?

- https://github.com/darktable-org/darktable/pull/13432

**Comment by garrett** [comment]
> Yet, there is probably a bug somewhere as we do allow now multiple lens instance.

Oh, does this mean that darktable no longer has a way to enable auto lens correction by default?

I really wish these corrections would be the default, without having to mess with presets or manually click the module for every photo. Same for noise reduction.

(Apologies for being offtopic.)

**Comment by TurboGit** [comment]
> Oh, does this mean that darktable no longer has a way to enable auto lens correction by default?

No it has. You could even have multiple instance auto-applied if needed.

**Comment by SoupyGit** [comment]
On the matter of allowing 2 lens instances, it is probably for the use case of digital negatives? One instance to correct the lens photographing the film, another to correct the lens photographing the scene.

**Comment by jenshannoschwalm** [comment]
you are probably right on this, that would make sense.

**Comment by TurboGit** [comment]
@garrett : I don't see **one** auto-applied lens preset by **seven**. So it is expected that you can have multiple lens correction applied. And as explained above it was by chance working before as only one auto-applied preset was selected. Now all the matching ones are. I bet that the issue you have is with a camera/lens matching multiple entries listed above in your screenshot.

At this point, I don't see a bug in dt. Correct me if I'm wrong.

**Comment by TurboGit** [comment]
> So, now, I have a choice between no lens profile or two copies of lens profiles double-applying corrections? thinking (Unless there's a non-obvious way (to me, at least) to set up a preset. Or if darktable auto-enables it in the future, like it does for other basic modules.)

No, the solution is to ensure that your presets are all exclusive. One solution (looking at your screenshot) is to have only the "Auto Lens Correction" set to auto applied as it matches all camera/lens/focal and remove the auto-applied flags for all other presets.

**Comment by garrett** [comment]
> I don't see one auto-applied lens preset by seven

No, there's just one preset for auto-matching.

The others are for lenses that darktable doesn't recognize in the metadata for whatever reason, so I can manually select lenses I've used.

For example: My very old Canon D60 (that I no longer have) doesn't seem to have its lens matched by darktable. I usually restrict them by camera and focal length (which does always show up, even if the lens itself isn't recognized) so I don't have a mess of options that don't matter to the image; only ones that match lenses I've actually had and used.

The bug isn't about manual presets; it's about what's being applied by default, even on cameras and lenses that I don't have (from Pixls.us).

**Comment by TurboGit** [comment]
> No, there's just one preset for auto-matching.

So I have no idea why you have multiple lens module applied.

**Comment by garrett** [comment]
> So I have no idea why you have multiple lens module applied.

Yeah, that's the bug. I wasn't able to figure it out either. But it's something to do with my settings, I guess.

I've had settings across many versions of darktable for many years. Perhaps it's time to rebuild it from scratch. :frowning_face: I'm just worried that others will hit this same bug, or that I will again after redoing everything. If it's just an isolated incident based on my own settings, then it's a bit different. I can at least try with a temporary subset and see if it happens again.

However, since the only way one can automatically apply lens profiles is via a preset, which is what I'm doing, I'm worried that this might not be an isolated experience as darktable has changed how it acts with regard to this module. I'll try to figure it out this evening or weekend (whenever I get the time).

**Comment by TurboGit** [comment]
I have looked at the RAW + XMP posted and again I don't see a bug but clearly two matching presets.

The RAW is from a **Fujifim X-H1** and a **18-55** lens.

In the screenshot about you have an auto-applied preset for this camera lens named `Fuji 18-55` and of course the generic `Auto Lens Correction` matching everything. So you end up with two lens correction module, no magic :)

This is what I also see in the XMP, confirming that both presets have been applied (see `multi_name`):

```
     <rdf:li
      darktable:num="8"
      darktable:operation="lens"
      darktable:enabled="1"
      darktable:modversion="7"
      darktable:params="gz05eJxjYGBgYGeAgEmODfaNUw/bMzBMcARiBwaGKhdGoHiErochwwCCCDdDC11T09xcNyM9C10ThSAFH18Ff89g+rqiwR4Z371WZQ8AhtESZQ=="
      darktable:multi_name="Auto Lens Correction"
      darktable:multi_name_hand_edited="0"
      darktable:multi_priority="1"
      darktable:blendop_version="12"
      darktable:blendop_params="gz13eJxjYGBgYAZiCQYYOOHEgAYY0QVwggZ7CB6pfNoAAE4AGQc="/>

     <rdf:li
      darktable:num="10"
      darktable:operation="lens"
      darktable:enabled="1"
      darktable:modversion="7"
      darktable:params="gz05eJxjYGBgYGeAgEmODfaNUw/bMzBMcARiBwaGKhdGoHiErochwwCCCDdDC11T09xcNyM9C10ThSAFH18Ff89g+rqiwR4Z371WZQ8AhtESZQ=="
      darktable:multi_name="Fuji 18-55"
      darktable:multi_name_hand_edited="0"
      darktable:multi_priority="0"
      darktable:blendop_version="12"
      darktable:blendop_params="gz13eJxjYGBgYAZiCQYYOOHEgAYY0QVwggZ7CB6pfNoAAE4AGQc="/>
```
